### DISCUSSION TITLE
github auth with sveltekit: Works in local dev, but prod rollout in Cloudflare Pages returns CallbackRouteError

### DISCUSSION THREAD

**Discussion by kpbaseten** [nextauthjs/next-auth#6653] [Help] [answered]
Trying to get github authentication working in a sveltekit site, and everything works well in a local dev environment, but when deploying into a Cloudflare Pages for production, the callback from github does not seem to work. Get a "error=CallbackRouteError".

I do have AUTH_TRUST_HOST="true", as well as trustHost: "true" in hooks just in case. Debug enabled, but unfortunately not able to get get logs (given it's a cloudflare pages site).

The callback url configured is http://localhost:1027/auth/callback/github for the local dev environment, and https://my.domain.com/auth/callback/github for the cloudflare pages site.

Any suggestions or clues on where to look?

**Comment by jschlesser** [comment]
Did you get it to work with local dev using vite (sveltekit) or using wrangler with the local option?   Ive gotten the sveltekit vite based environment to work, which is no surprise as it's a node api environment.   wrangler / cloudflare is a javascript worker environment so not everything works the same.  When I try it locally in wrangler I get the following error `Cannot read properties of undefined (reading 'get')` on this line of code `pragma = res.headers.get("cache-control")`, which strongly implies some kind of underlying request api issue.

**Reply by kpbaseten** [reply]
Thanks for the insight. I got it to work locally with vite.

Couldn't quite decipher your other comment - are you suggesting that it's a request api issue in auth.js, or in cloudflare pages/wrangler?

**Reply by jschlesser** [reply]
EDIT:  This wasn't it
My comment about differences between node and wrangler / cloudflare has to do with how they implement outgoing http requests.  Specifically the fetch api.  wrangler under the hood is a node application but it uses miniflare to mock the cloudflare platform features and miniflare uses undici for its fetch implementation.

Turns out the error I was getting locally was just my code, after fixing that, I get the same callback route error you get when running in wrangler.   After turning all the debugging on that I could in both wrangler and authjs, I get an additional bit of error logging.   It could be that it is a network based error.  My understanding is that when the callback URL is invoked in an OAuth2 workflow and the user is authenticated, then the callback needs to exchange the authorization token in the callback for an access token and a refresh token. [See OAuth2 workflow](https://www.c-sharpcorner.com/article/understanding-workflow-of-oauth2-0-authorization-grant-types/)  That exchange is server to server and I think that is what is failing.  There isn't enough debugging data to be 100% sure though. 

`[auth][cause]: TypeError: terminated
    at Fetch.onAborted (/opt/homebrew/lib/node_modules/wrangler/node_modules/undici/lib/fetch/index.js:1890:49)
    at Fetch.emit (node:events:513:28)
    at Fetch.emit (node:domain:489:12)
    at Fetch.terminate (/opt/homebrew/lib/node_modules/wrangler/node_modules/undici/lib/fetch/index.js:83:10)
    at Object.onError (/opt/homebrew/lib/node_modules/wrangler/node_modules/undici/lib/fetch/index.js:2027:34)
    at Request.onError (/opt/homebrew/lib/node_modules/wrangler/node_modules/undici/lib/core/request.js:265:27)
    at errorRequest (/opt/homebrew/lib/node_modules/wrangler/node_modules/undici/lib/client.js:1722:13)
    at TLSSocket.onSocketClose (/opt/homebrew/lib/node_modules/wrangler/node_modules/undici/lib/client.js:985:5)
    at TLSSocket.emit (node:events:525:35)
    at TLSSocket.emit (node:domain:489:12)
[auth][details]: {
  "name": "SocketError",
  "code": "UND_ERR_SOCKET",
  "socket": {
    "bytesWritten": 180,
    "bytesRead": 556
  },
  "provider": "github"
}`

**Reply by jschlesser** [reply]
Edit: And this wasn't it either
Here is a bit more evidence to support my hypothesis.
A [bug](https://github.com/nodejs/undici/issues/1670) in the indici module about accessing GitHub apis and getting UND_ERR_SOCKET.
Note that the bug was fixed and [merged](https://github.com/nodejs/undici/pull/1726) on Oct 23, 2022

However miniflare currently [uses](https://github.com/cloudflare/miniflare/blob/ae477585b9abea93a5f7c49a4709ac129ebb7141/package-lock.json#L9610) indici 5.11.0 

And indici [5.11.0](https://github.com/nodejs/undici/releases/tag/v5.11.0) was released on Oct 3, 2022 (20 days before the fix)

**Reply by jschlesser** [reply]
I finally figured this out with the help of running a proxy to capture outbound traffic.   Turns out it isn't indici at all.   The GitHub rest api requires the user-agent to be set during requests.  Much of the GitHub providers requests go through the default / built in OAuth library, however the GitHub specific provider code uses some fetch calls to get the user profile information (like email) and that code isn't setting a user-agent and the version of indici that is running on wrangler doesnt provide a generic one, it just omits it, which is probably a good behavior from a security perspective anyhow.   I have a local patched version of authjs GitHub provider that works in wrangler pages dev locally.  If you add this into the headers of the GitHub provider node_modules/@auth/core/providers/github.js your local code and then rebuild your app and run it in wrangler pages dev, it will work.   `, 'User-Agent': 'undici'`

**Reply by jschlesser** [reply]
Im working on a PR for this as well as a PR for some changes to the sveltekit framework in authjs so that it can get environment variables from event.platform.env which is the only way to get environment variables into svelte on cloudflare workers / pages.  Everybody will need that for setting the client_id and secret.   Assuming the PR's get accepted ill post a cloudflare pages specific template online for the community.

**Comment by jschlesser** [comment, marked as answer]
Quick update on some Issues and PRs to track this in case anybody else makes it here.
user-agent [issue](https://github.com/nextauthjs/next-auth/issues/6741)
environment variables [issue](https://github.com/nextauthjs/next-auth/issues/6743)

Note that this issue is only if you are trying to build worker.js for wrangler inside of the next_auth app/dev/sveltekit environment
Building with pnpm in dev [issue](https://github.com/nextauthjs/next-auth/issues/6732)

**Reply by kpbaseten** [reply]
Amazing work on root-causing the issue and building a fix jschlesser , and the [github docs](https://github.com/nextauthjs/next-auth/pull/6742#pullrequestreview-1303203161) seem to be clear on the requirement for the User-agent header.

So looks like the PR is just awaiting an additional review and merge by ThangHuuVu

**Reply by jschlesser** [reply]
The user-agent fix has been merged into main.

**Reply by kpbaseten** [reply]
Thanks for the fix, @jschlesser. Now I no longer get a CallbackRouteError, but now the issue instead is that after successful authentication I get redirected back to the auth page.

**Comment by jschlesser** [comment]
First make sure you are on the latest @auth/core@0.5.1  and @auth/sveltekit@0.3.0, you must be on the core version since that has the user agent header fix for GitHub.

After that, the standard code example works fine for me.   It looks like your hooks code has an authorization function that isn't in the standard example in the repository.  It looks like it is using the session as well, id make sure that you have some kind of session support set up, that is either going to be part of the adapter you are using or using JWT.  Looking at your code in the hooks file, my guess is that your issue has to do with session somehow.

**Comment by jschlesser** [comment]
Also, technically that is a different issue.   It would be great if you could mark this one as answered.

**Comment by kpbaseten** [comment]
@jschlesser _finally_ got it working! It seems there's quite a few quirks to the auth.js + github + sveltekit + cloudflare combination. Your PR's were instrumental to fixing the auth.js w/ github on CF issues. Needed a few more cf + sveltekit specific workarounds to get auth.js to work with github, including turn off prerender (!), fixing js code where github-required headers were missing, etc.

Thanks a heap for your work on the fixes!
